### PR TITLE
CPP: Improve locations for AV Rule 114.ql.

### PR DESCRIPTION
**Pull request by geoffw0**
I've been meaning to fix this for a while.  We try to be clever about where we flag the location for this query, but sometimes we just end up confusing things.

Differences query: https://lgtm.com/query/2894440083018682314/

### CODE REVIEW

File: cpp/ql/test/query-tests/jsf/4.13 Functions/AV Rule 114/test.cpp [hunk 1]
```diff
@@ -48,7 +48,7 @@ MyValue g7(bool c)
 	if (c) return MyValue(7);
 	DONOTHING
 	DONOTHING
-	// BAD [the alert here is unfortunately placed]
+	// GOOD
```
Comment (rdmarsh2) [inline cpp/ql/test/query-tests/jsf/4.13 Functions/AV Rule 114/test.cpp]: Shouldn't this be `// BAD` - i.e. it is correct to produce an alert here?

Comment (geoffw0) [reply cpp/ql/test/query-tests/jsf/4.13 Functions/AV Rule 114/test.cpp]: Good point.  Corrected.

**Comment by geoffw0** [comment]
Updated and added a change note.

**Comment by rdmarsh2** [comment]
It looks like the change note now has a conflict - I'm happy to merge once that's addressed.

**Comment by geoffw0** [comment]
Fixed, sorry for the delay.

**Comment by geoffw0** [comment]
Fixed merge conflict again.

File: cpp/ql/src/jsf/4.13 Functions/AV Rule 114.ql [hunk 2]
```diff
@@ -30,7 +30,13 @@ predicate functionsMissingReturnStmt(Function f, ControlFlowNode blame) {
   ) and
   exists(ReturnStmt s |
     f.getAPredecessor() = s and
-    blame = s.getAPredecessor()
+    (
+      blame = s.getAPredecessor() and
+      count(blame.getASuccessor()) = 1
+      or
```
Comment (zlaski) [inline cpp/ql/src/jsf/4.13 Functions/AV Rule 114.ql]: Under what circumstances would `blame != s`?

Comment (geoffw0) [reply cpp/ql/src/jsf/4.13 Functions/AV Rule 114.ql]: Consider something like this:
```
int myFunc(bool cond)
{
	if (cond)
	{
		doSomething();
	} else {
		doSomethingElse();
	}
	// there's an implicit `return;` here, called `s` in the QL.
}
```
In this case `s` has two predecessors, the calls to `doSomething` and `doSomethingElse`.  Selecting either or both of them is confusing, as it suggests there is something wrong with those calls when in fact the issue is a missing return statement several lines below them.

On the flip side for a simple case:
```
int myFunc(bool cond)
{
	doSomething();
}
```
it was found that selecting `doSomething` is clearer than selecting the closing `}` (this is open to discussion; I don't think anything breaks if we just select the closing `}` in all cases, for simplicity and consistency, but the original query was deliberately designed to select `doSomething` in this case)

Comment (geoffw0) [reply cpp/ql/src/jsf/4.13 Functions/AV Rule 114.ql]: The example in test `g7` is similar but even worse, due to the presence of macros that look like they are code (this is based on something we found in the wild).

**Comment by geoffw0** [comment]
Rebased and moved change note onto 1.24.

**Comment by geoffw0** [comment]
@rdmarsh2 this has been floating around for a while, please could you merge it?

**Comment by geoffw0** [comment]
Fixed merge conflict.

**Pull request merged**